allennlp/models/archival.py

```diff
@@ -158,43 +157,40 @@ def load_archive(
-        serialization_dir = resolved_archive_file
-    else:
-        # Extract archive to temp dir
-        tempdir = tempfile.mkdtemp()
-        logger.info(f"extracting archive file {resolved_archive_file} to temp dir {tempdir}")
-        with tarfile.open(resolved_archive_file, "r:gz") as archive:
-            archive.extractall(tempdir)
-        # Postpone cleanup until exit in case the unarchived contents are needed outside
-        # this function.
-        atexit.register(_cleanup_archive_dir, tempdir)
-
-        serialization_dir = tempdir
-
-    # Load config
-    config = Params.from_file(os.path.join(serialization_dir, CONFIG_NAME), overrides)
-
-    if weights_file:
-        weights_path = weights_file
-    else:
-        weights_path = os.path.join(serialization_dir, _WEIGHTS_NAME)
-        # Fallback for serialization directories.
-        if not os.path.exists(weights_path):
-            weights_path = os.path.join(serialization_dir, _DEFAULT_WEIGHTS)
-
-    # Instantiate model. Use a duplicate of the config, as it will get consumed.
-    model = Model.load(
-        config.duplicate(),
-        weights_file=weights_path,
-        serialization_dir=serialization_dir,
-        cuda_device=cuda_device,
-    )
-
-    return Archive(model=model, config=config)
-
-
-def _cleanup_archive_dir(path: str):
-    if os.path.exists(path):
-        logger.info("removing temporary unarchived model dir at %s", path)
-        shutil.rmtree(path)
+    tempdir = None
+    try:
+        if os.path.isdir(resolved_archive_file):
+            serialization_dir = resolved_archive_file
+        else:
+            # Extract archive to temp dir
+            tempdir = tempfile.mkdtemp()
+            logger.info(f"extracting archive file {resolved_archive_file} to temp dir {tempdir}")
+            with tarfile.open(resolved_archive_file, "r:gz") as archive:
+                archive.extractall(tempdir)
+            serialization_dir = tempdir
+
+        # Load config
+        config = Params.from_file(os.path.join(serialization_dir, CONFIG_NAME), overrides)
+
+        if weights_file:
+            weights_path = weights_file
+        else:
+            weights_path = os.path.join(serialization_dir, _WEIGHTS_NAME)
+            # Fallback for serialization directories.
+            if not os.path.exists(weights_path):
+                weights_path = os.path.join(serialization_dir, _DEFAULT_WEIGHTS)
+
+        # Instantiate model. Use a duplicate of the config, as it will get consumed.
+        model = Model.load(
+            config.duplicate(),
+            weights_file=weights_path,
+            serialization_dir=serialization_dir,
+            cuda_device=cuda_device,
+        )
+
+        return Archive(model=model, config=config)
+
+    finally:
+        if tempdir is not None:
+            logger.info(f"removing temporary unarchived model dir at {tempdir}")
+            shutil.rmtree(tempdir, ignore_errors=True)
```

Comment:
Does this work with the return statement above the `finally` block? I would think the `finally` block would be skipped if this method successfully returned.

Either way, I think it's more clear to have the return statement after the `finally` block.

Reply:
Sure, changed it.

Reply:
I think that technically whatever is in the `finally` block will be executed even when there is a `return` in the `try` block. This is according the [Python docs](https://docs.python.org/3/reference/compound_stmts.html#the-try-statement). But, in any case, it's clearer, as you mentioned.